spring-kafka/src/main/java/org/springframework/kafka/annotation/RetryableTopicAnnotationProcessor.java

```diff
@@ -386,4 +363,18 @@ private String resolve(String value) {
 		return value;
 	}
 
+	private static final class BackOffValues {
```

Comment:
This looks like Java `record`: https://blogs.oracle.com/javamagazine/post/records-come-to-java.

Although I wonder why do we need this intermediate class at all?
Why plain `HashMap` doesn't work?
Because of type?
But this is an internal logic, so simple casting won't hurt and won't affect anything else.

Reply:
Hi @artembilan, thanks for looking into this. I've been meaning to ask about java 17 vs java 8 - but I guess the real question is: what's the criteria for a feature to be cherry-picked to 2.8.x, or only 3.0.0? I've been sticking with Java 8 since 3.0.0 seems to be a while away.

As for `HashMap`, correct me if I'm wrong, but I think it's more a matter of personal preference, or maybe framework standards. I for one tend to think that having a small inner class, or record, is easier to reason about than a Map, since the compiler helps us in not messing anything up, and the properties can have nice descriptive names. Are there any other tradeoffs I might be missing? Of course, I can change it to a `Map` if that's better, no worries.

I also thought of having this values class as an inner class of `BackOffProvider` for client classes to use, but then the class would have to be public, which seems like an unnecessary exposure since it's not meant for end users.

Reply:
OK. You know it sounds more like a `BackOffPolicyBuilder` has to be there.
Perhaps for now it could be implemented here in `spring-kafka`, but more logical to have it in the `spring-retry` in that `org.springframework.retry.backoff` package.
I see there is some API on the matter in the `RetryTemplateBuilder`:
```
public RetryTemplateBuilder exponentialBackoff(long initialInterval, double multiplier, long maxInterval,
			boolean withRandom) {

public RetryTemplateBuilder fixedBackoff(long interval) {

public RetryTemplateBuilder uniformRandomBackoff(long minInterval, long maxInterval) {

public RetryTemplateBuilder noBackoff() {

public RetryTemplateBuilder customBackoff(BackOffPolicy backOffPolicy) {
```

Doesn't that look we are duplicating over here in SK?

Reply:
Sure, maybe we can leave it here for now, open a PR for Spring Retry for having a `BackOffPolicyBuilder`, and when we have the API exposed we can use it and remove from here. We don't use the actual RetryTemplate for the Retry Topics feature, only the `BackOffPolicy` and `@BackOff` annotation. You can see how the policy is used in the [BackOffValuesGenerator](https://github.com/spring-projects/spring-kafka/blob/bee97d9c8fcc1f9489b4244a653ef19b4f8abc83/spring-kafka/src/main/java/org/springframework/kafka/retrytopic/BackOffValuesGenerator.java) class.

Also, instead of having the inner class, we could pass the builder as an argument to the `createBackoffValues` method and just set it there, but then we'd need to break the fluent API to get a reference to the builder - which doesn't look like a problem in an internal code, as you mentioned.

When this feature was developed the Spring Retry project didn't seem to be getting many releases, so it didn't seem really practical to wait for it instead of simply duplicating the logic, since it's not that much code. I see the Spring team made a task force to make a new release of the project, thanks and congrats for that.

Reply:
Well, you are right: we are committed to have a better support for Spring Retry.
So, if you would implement a `BackOffPolicyBuilder`, then it is better, please, to PR it against `spring-retry`.
Then the current `RetryTopicConfigurationBuilder` would not have any changes.
And then we would benefit in Spring Boot as well.
We won't have that `BackOffValues` and so on.

Does it make sense?

Reply:
I haven't looked at this yet, but if it's not too complex, I don't mind putting it into 2.8.x because 3.0.0 is many months away.

However, if we go ahead and do the "retry topics all in the main container" enhancement, I would start a 2.9 soon (for Boot 2.7) and we can stop back-porting new features like this.

Reply:
> Well, you are right: we are committed to have a better support for Spring Retry. So, if you would implement a `BackOffPolicyBuilder`, then it is better, please, to PR it against `spring-retry`. Then the current `RetryTopicConfigurationBuilder` would not have any changes. And then we would benefit in Spring Boot as well. We won't have that `BackOffValues` and so on.
> 
> Does it make sense?

Sure. Let me just make sure we're on the same page. There are currently two ways of getting a `BackOffPolicy`, both in Spring Kafka and Spring Retry.

The first is through a `Builder` where we explicitly say the policy we want - as in the methods you mentioned. This part doesn't seem like something we'd change - it's already part of the public APIs of both projects.

The second is inferring the policy through the values we pass, which is used in the `@BackOff` annotation processing, originally in the [AnnotationAwareRetryOperationsInterceptor](https://github.com/spring-projects/spring-retry/blob/302f7999932421d39800b81b5fa9f6b1c557b957/src/main/java/org/springframework/retry/annotation/AnnotationAwareRetryOperationsInterceptor.java#L404) class in the `Spring Retry` project, and duplicated in the [RetryableTopicAnnotationProcessor.java](https://github.com/spring-projects/spring-kafka/blob/43cb78127b3851c03037418533bf393c2fc7db3a/spring-kafka/src/main/java/org/springframework/kafka/annotation/RetryableTopicAnnotationProcessor.java#L171) class, which is the  logic I extracted for the `BackOffProvider` class for this PR, and the one we need for Boot's auto configuration. This sounds like the one we should open up and deduplicate.

So my question is, are you proposing a `BackOffPolicyBuilder` that has both approaches, e.g., a class with the `Builder` methods you mentioned, and also a generic `long, long, double, boolean` method for inferring the policy?

Reply:
I think we just expose simple props and have a complex logic in the `build()` to determine what impl to return.

I guess we can start from a simple builder approach to satisfy our current requirements.
It could evolve eventually.

Reply:
Oh, now I see. You mean having something like:

```
BackOffPolicyBuilder
    .withValues()
    .minDelay(1000L)
    .maxDelay(5000L)
    .multiplier(3.0)
    .random()
    .build()
```

That sure looks much better than what I suggested, thanks. If that's about what you had in mind, maybe we can close this PR and I'll open a new one for Spring Retry soon.

Reply:
Something like that, yes.
But probably without `.withValues()` and `random()` with param. We got that `true/false` from the annotation or Spring Boot configuration properties.
I believe it would look better to have it like this: `random(randomValue)` rather than `if..else`.

Reply:
Sure, makes sense, thanks. I'll open a PR there then, and we can improve the API as needed. I'll also take a look into how Spring Retry's builders are implemented.